pseudopotential_calculator/calculations/adsorbate.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import matplotlib.pyplot as plt
+import numpy as np
+from ase import Atom, Atoms
+from matplotlib import cm
+
+from pseudopotential_calculator.atoms import Vector, get_basis_vectors, repeat_cell
+from pseudopotential_calculator.calculations.slab import get_height_of_layer
+from pseudopotential_calculator.castep import (
+    get_calculator_atom,
+)
+
+if TYPE_CHECKING:
+    from ase.calculators.castep import Castep
+    from matplotlib.figure import Figure
+
+
+def get_top_position(atoms: Atoms) -> tuple[float, float, float]:
+    positions = atoms.positions  # type: ignore bad lib
+    return tuple(max(positions, key=lambda x: x[2]))  # type: ignore bad lib
+
+
+def get_top_layer_z_vector(atoms: Atoms) -> tuple[float, float, float]:
+    z = get_height_of_layer(atoms, 0)
+    return (0, 0, z)
+
+
+def get_adsorbate_position(
+    slab_position: tuple[float, float, float],
+    relative_height: float | None = None,
+) -> tuple[float, float, float]:
+    if relative_height is None:
+        relative_height = 1.0
+    return (
+        slab_position[0],
+        slab_position[1],
+        slab_position[2] + relative_height,
+    )
+
+
+def prepare_adsorbate(
+    adsorbate: str,
+    position: tuple[float, float, float],
+) -> Atom:
+    return Atom(adsorbate, position)
+
+
+def get_slab_with_adsorbate(atom: Atom, slab: Atoms) -> Atoms:
+    slab.append(atom)  # type: ignore bad lib)
+    return slab
```

Comment:
this is confusing - i would expect if it is returning a new slab it would not modify the old one - either remove the return statement or clone the old one before appending